flexmeasures/data/services/time_series.py

```diff
@@ -329,6 +329,7 @@ def drop_unchanged_beliefs(bdf: tb.BeliefsDataFrame) -> tb.BeliefsDataFrame:
         event_ends_before=bdf.event_ends[-1],
         beliefs_before=bdf.lineage.belief_times[0],  # unique belief time
         source=bdf.lineage.sources[0],  # unique source
+        most_recent_beliefs_only=False,
```

Comment:
Right after this, we select the most recent from `previous_beliefs_in_db`. Why not just call `search_beliefs` with `most_recent_beliefs_only=True`, then?

Reply:
It turns out that the `most_recent_beliefs` and `most_recent_events` subqueries in timely-beliefs are missing the optional filters that are also applied to the main query.

Unfortunately, I also found an incompatibility. Two of those filters only roughly filter within the query, and only properly filter on the query results: namely `beliefs_after` and `beliefs_before`. They do so because they require taking into account the sensor's knowledge horizon function, which are python functions and not suitable for injection into the SQL query. The SQL filters essentially introduce some slack around the restricted range of belief times, with the size of this slack equal to the bounds of the specific knowledge horizon function. For example, the knowledge horizon for events of a day-ahead market with a gate closure at noon is normally bounded between 12 and 36 hours (to account for possibly double daylight savings time transitions this is even widened to between 10 and 38 hours). In any case, this messes up selecting the minimum horizon in the subquery, especially for economic sensors with non-trivial knowledge horizons (physical sensors are more straightforward and typically lead to no slack being introduced).

I suggest I follow up with a timely-beliefs issue, and keep these FlexMeasures code lines as is for now.

Reply:
Complicated... Maybe add a comment here, at least.

Reply:
I opened https://github.com/SeitaBV/timely-beliefs/issues/97 and added a comment.